cli/sdo/voucher.go

```diff
@@ -0,0 +1,310 @@
+		if err == io.EOF {
+			break	// this means we hit the end of the tar file
+		} else if err != nil {
+			cliutils.Fatal(cliutils.CLI_INPUT_ERROR, msgPrinter.Sprintf("reading the bytes from %s: %v", voucherFileName, err))
+		}
+		//fmt.Printf("file name: %s\n", header.Name)
+		switch header.Typeflag {
+		case tar.TypeDir:
+			continue // just ignore the directories
+		case tar.TypeReg:
+			// Regular file, only process it if it's a json file
+			if strings.HasPrefix(header.Name, ".") || !strings.HasSuffix(header.Name, ".json") {
+				continue
+			}
+			import1Voucher(org, userCreds, sdoUrl, tarReader, header.Name, example, policyFilePath, true)
+		}
+	}
+}
+
+func importZip(org, userCreds, sdoUrl string, voucherFile *os.File, example, policyFilePath string) {
+	msgPrinter := i18n.GetMessagePrinter()
+	voucherBytes, err := ioutil.ReadAll(bufio.NewReader(voucherFile))
+	if err != nil {
+		cliutils.Fatal(cliutils.CLI_INPUT_ERROR, msgPrinter.Sprintf("reading the bytes from %s: %v", voucherFile.Name(), err))
+	}
+	zipReader, err := zip.NewReader(bytes.NewReader(voucherBytes), int64(len(voucherBytes)))
+	if err != nil {
+		cliutils.Fatal(cliutils.CLI_INPUT_ERROR, msgPrinter.Sprintf("creating zip reader for %s: %v", voucherFile.Name(), err))
+	}
+	for _, fileInfo := range zipReader.File {
+		if strings.HasPrefix(fileInfo.Name, ".") || !strings.HasSuffix(fileInfo.Name, ".json") {
+			continue
+		}
+		zipFileReader, err := fileInfo.Open()
+		if err != nil {
+			cliutils.Fatal(cliutils.CLI_INPUT_ERROR, msgPrinter.Sprintf("opening file %s within zip for %s: %v", fileInfo.Name, voucherFile.Name(), err))
+		}
+		import1Voucher(org, userCreds, sdoUrl, zipFileReader, fileInfo.Name, example, policyFilePath, true)
+		zipFileReader.Close()
+	}
+}
+
+func import1Voucher(org, userCreds, sdoUrl string, voucherFileReader io.Reader, voucherFileName, example, policyFilePath string, quieter bool) {
+	msgPrinter := i18n.GetMessagePrinter()
+
+	// Parse the voucher so we can tell them what we are doing
+	outStruct := InspectOutput{}
+	voucherBytes, err := ioutil.ReadAll(voucherFileReader)
+	if err != nil {
+		cliutils.Fatal(cliutils.CLI_INPUT_ERROR, msgPrinter.Sprintf("reading the bytes from %s: %v", voucherFileName, err))
+	}
+	if err = parseVoucherBytes(voucherBytes, &outStruct); err != nil {
+		cliutils.Fatal(cliutils.JSON_PARSING_ERROR, msgPrinter.Sprintf("parsing the json from %s: %v", voucherFileName, err))
+	}
+	// This is the single msg we want displayed even when quieter==true
+	msgPrinter.Printf("Importing %s for device %s, using rendezvous servers %s ...", voucherFileName, outStruct.Device.Uuid, strings.Join(outStruct.Voucher.RendezvousUrls, ", "))
+	msgPrinter.Println()
+
+	// Import the voucher to the SDO owner service
+	creds := cliutils.OrgAndCreds(org, userCreds)
+	importResponse := ImportResponse{}
+	SdoPostVoucher(sdoUrl+"/voucher", creds, voucherBytes, &importResponse)
```

Comment:
The path.Join function supports URL construction with proper slash handling so that we dont have to worry about trailing slashes on URL prefixes, like sdoUrl.

Reply:
Thx for the tip, fixed.